source/blog/2016-04-01-deadlock-investigation.html.md

```diff
@@ -0,0 +1,191 @@
+/opt/rubies/ruby-2.2.4/lib/ruby/gems/2.2.0/bundler/gems/rails-168836772753/activesupport/lib/active_support/dependencies.rb:511:in `load_missing_constant'
+/opt/rubies/ruby-2.2.4/lib/ruby/gems/2.2.0/bundler/gems/rails-168836772753/activesupport/lib/active_support/dependencies.rb:203:in `const_missing'
+/opt/rubies/ruby-2.2.4/lib/ruby/gems/2.2.0/bundler/gems/rails-168836772753/activesupport/lib/active_support/dependencies.rb:543:in `load_missing_constant'
+/opt/rubies/ruby-2.2.4/lib/ruby/gems/2.2.0/bundler/gems/rails-168836772753/activesupport/lib/active_support/dependencies.rb:203:in `const_missing'
+/opt/rubies/ruby-2.2.4/lib/ruby/gems/2.2.0/bundler/gems/rails-168836772753/activesupport/lib/active_support/dependencies.rb:543:in `load_missing_constant'
+/opt/rubies/ruby-2.2.4/lib/ruby/gems/2.2.0/bundler/gems/rails-168836772753/activesupport/lib/active_support/dependencies.rb:203:in `const_missing'
+/var/www/miq/vmdb/app/controllers/api_controller/renderer.rb:204:in `collection_search'
+```
+
+Note, the interlock.rb is used for autoloading constants.
+
+For example, to print the new rails 5 interlock
+
+```
+call (void)rb_eval_string("p ActiveSupport::Dependencies.interlock")
+```
+
+It looked something like this:
+```
+#<ActiveSupport::Dependencies::Interlock:0x0000000a24d1c8
+  @lock=#<ActiveSupport::Concurrency::ShareLock:0x0000000a24d178
+  @mon_owner=nil,
+  @mon_count=0,
+  @mon_mutex=#<Mutex:0x0000000a24d128>,
+  @cv=#<MonitorMixin::ConditionVariable:0x0000000a24d100
+  @monitor=#<ActiveSupport::Concurrency::ShareLock:0x0000000a24d178 ...>,
+  @cond=#<Thread::ConditionVariable:0x0000000a24d0d8>>,
+  @sharing={
+    #<Thread:0x00000007ae8c18@/opt/rubies/ruby-2.2.4/lib/ruby/gems/2.2.0/gems/puma-2.16.0/lib/puma/thread_pool.rb:61 dead>=>1,
+    #<Thread:0x00000007ae9230@/opt/rubies/ruby-2.2.4/lib/ruby/gems/2.2.0/gems/puma-2.16.0/lib/puma/thread_pool.rb:61 dead>=>1,
+    #<Thread:0x00000007ac9f48@/opt/rubies/ruby-2.2.4/lib/ruby/gems/2.2.0/gems/puma-2.16.0/lib/puma/thread_pool.rb:61 dead>=>1},
+  @waiting={
+    #<Thread:0x000000025420e8@/opt/rubies/ruby-2.2.4/lib/ruby/gems/2.2.0/gems/puma-2.16.0/lib/puma/thread_pool.rb:61 sleep>=>[:load, [:load]],
+    #<Thread:0x00000007c6bc48@/opt/rubies/ruby-2.2.4/lib/ruby/gems/2.2.0/gems/puma-2.16.0/lib/puma/thread_pool.rb:61 sleep>=>[:load, [:load]],
+    #<Thread:0x00000007c6b810@/opt/rubies/ruby-2.2.4/lib/ruby/gems/2.2.0/gems/puma-2.16.0/lib/puma/thread_pool.rb:61 sleep>=>[:load, [:load]]}, @exclusive_thread=nil, @exclusive_depth=0>>
+```
+
+Note, if you look at the threads in the `@sharing` instance variable, you'll see they're dead threads and they're holding a lock.
+
+With this information, we instrument the [ActionDispatch::Executor](https://github.com/rails/rails/blob/7b82e1c77b48cb351da4e0ed6ea0bac806d4925c/actionpack/lib/action_dispatch/middleware/executor.rb#L9-L17),
+which handles acquiring a lock in the run! method and returning the lock in the complete! method.
```

Comment:
s/instrument/look to/ (the instrumentation comes in the next paragraph)

+ "This is the method responsible for ensuring acquired locks are always returned; if the thread is dying still holding a lock, the problem must lie somewhere nearby."